### ISSUE TITLE
ImportError: OpenCV loader: missing configuration file: ['config.py'].

### ISSUE THREAD

**Issue by clouds1203** [opencv/opencv-python#569]
### Expected behaviour

Using `pyinstaller -F MyProgram.py` to build exe, but it got error.

### Actual behaviour

I write my program with cv2, and it run well,
but after I build it into exe file by `pyinstaller -F MyProgram.py` it can't run.
below is error log:

Traceback (most recent call last):
  File "AutomaticTestTool.py", line 12, in <module>
  File "PyInstaller\loader\pyimod03_importers.py", line 546, in exec_module
  File "cv2\__init__.py", line 180, in <module>
  File "cv2\__init__.py", line 110, in bootstrap
  File "cv2\__init__.py", line 108, in load_first_config
ImportError: OpenCV loader: missing configuration file: ['config.py']. Check OpenCV installation.
[12104] Failed to execute script 'AutomaticTestTool' due to unhandled exception!

I have tried: 
`pip3 install --upgrade opencv-python
pip3 install --upgrade pyinstaller`

but it no use.

and below is log of pyinstaller :

pyinstaller AutomaticTestTool.py
112 INFO: PyInstaller: 4.5.1
112 INFO: Python: 3.9.1
112 INFO: Platform: Windows-10-10.0.19041-SP0
126 INFO: wrote C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\AutomaticTestTool.spec
127 INFO: UPX is not available.
130 INFO: Extending PYTHONPATH with paths
['C:\\Users\\WNC.T1-1-1-Q-52026\\Desktop\\1018',
 'C:\\Users\\WNC.T1-1-1-Q-52026\\Desktop\\1018']
456 INFO: checking Analysis
456 INFO: Building Analysis because Analysis-00.toc is non existent
456 INFO: Initializing module dependency graph...
456 INFO: Caching module graph hooks...
472 INFO: Analyzing base_library.zip ...
2559 INFO: Processing pre-find module path hook distutils from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-distutils.py'.
2559 INFO: distutils: retargeting to non-venv dir 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib'
4096 INFO: Caching module dependency graph...
4309 INFO: running Analysis Analysis-00.toc
4321 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\python.exe
4353 WARNING: lib not found: api-ms-win-core-path-l1-1-0.dll dependency of c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\python39.dll
4369 INFO: Analyzing C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\AutomaticTestTool.py
4480 INFO: Processing pre-safe import module hook urllib3.packages.six.moves from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-urllib3.packages.six.moves.py'.
5703 INFO: Processing pre-find module path hook site from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-site.py'.
5718 INFO: site: retargeting to fake-dir 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\fake-modules'
6416 INFO: Processing pre-safe import module hook setuptools.extern.six.moves from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-setuptools.extern.six.moves.py'.
8234 INFO: Processing module hooks...
8234 INFO: Loading module hook 'hook-certifi.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
8234 INFO: Loading module hook 'hook-cv2.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
8249 INFO: Loading module hook 'hook-difflib.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8254 INFO: Loading module hook 'hook-distutils.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8254 INFO: Loading module hook 'hook-distutils.util.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8258 INFO: Loading module hook 'hook-encodings.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8405 INFO: Loading module hook 'hook-heapq.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8405 INFO: Loading module hook 'hook-lib2to3.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8444 INFO: Loading module hook 'hook-multiprocessing.util.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8460 INFO: Loading module hook 'hook-numpy.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8507 INFO: Import to be excluded not found: 'f2py'
8522 INFO: Loading module hook 'hook-numpy._pytesttester.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8527 INFO: Loading module hook 'hook-pickle.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8530 INFO: Loading module hook 'hook-pkg_resources.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
8983 INFO: Processing pre-safe import module hook win32com from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\pre_safe_import_module\\hook-win32com.py'.
9071 WARNING: Hidden import "pkg_resources.py2_warn" not found!
9071 WARNING: Hidden import "pkg_resources.markers" not found!
9087 INFO: Loading module hook 'hook-setuptools.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\lib\site-packages\setuptools\distutils_patch.py:25: UserWarning: Distutils was imported before Setuptools. This usage is discouraged and may exhibit undesirable behaviors or errors. Please use Setuptools' objects directly or at least import Setuptools first.
  warnings.warn(
9641 INFO: Loading module hook 'hook-sysconfig.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
9641 INFO: Loading module hook 'hook-win32ctypes.core.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
9894 INFO: Loading module hook 'hook-xml.dom.domreg.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
9894 INFO: Loading module hook 'hook-xml.etree.cElementTree.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
9909 INFO: Loading module hook 'hook-xml.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
9909 INFO: Loading module hook 'hook-_tkinter.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
10149 INFO: checking Tree
10149 INFO: Building Tree because Tree-00.toc is non existent
10149 INFO: Building Tree Tree-00.toc
10228 INFO: checking Tree
10228 INFO: Building Tree because Tree-01.toc is non existent
10228 INFO: Building Tree Tree-01.toc
10396 INFO: checking Tree
10396 INFO: Building Tree because Tree-02.toc is non existent
10396 INFO: Building Tree Tree-02.toc
10396 INFO: Loading module hook 'hook-setuptools.msvc.py' from 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
10428 INFO: Looking for ctypes DLLs
10474 INFO: Analyzing run-time hooks ...
10474 INFO: Including run-time hook 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgutil.py'
10474 INFO: Including run-time hook 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_multiprocessing.py'
10490 INFO: Including run-time hook 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_inspect.py'
10495 INFO: Including run-time hook 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth__tkinter.py'
10498 INFO: Including run-time hook 'c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\rthooks\\pyi_rth_certifi.py'
10500 INFO: Looking for dynamic libraries
10920 INFO: Looking for eggs
10920 INFO: Using Python library c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\python39.dll
10920 INFO: Found binding redirects:
[]
10920 INFO: Warnings written to C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\build\AutomaticTestTool\warn-AutomaticTestTool.txt
10983 INFO: Graph cross-reference written to C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\build\AutomaticTestTool\xref-AutomaticTestTool.html
11034 INFO: checking PYZ
11034 INFO: Building PYZ because PYZ-00.toc is non existent
11034 INFO: Building PYZ (ZlibArchive) C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\build\AutomaticTestTool\PYZ-00.pyz
11625 INFO: Building PYZ (ZlibArchive) C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\build\AutomaticTestTool\PYZ-00.pyz completed successfully.
11650 INFO: checking PKG
11650 INFO: Building PKG because PKG-00.toc is non existent
11650 INFO: Building PKG (CArchive) PKG-00.pkg
11690 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
11703 INFO: Bootloader c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\lib\site-packages\PyInstaller\bootloader\Windows-64bit\run.exe
11703 INFO: checking EXE
11703 INFO: Building EXE because EXE-00.toc is non existent
11703 INFO: Building EXE from EXE-00.toc
11742 INFO: Copying icons from ['c:\\users\\wnc.t1-1-1-q-52026\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\bootloader\\images\\icon-console.ico']
11744 INFO: Writing RT_GROUP_ICON 0 resource with 104 bytes
11744 INFO: Writing RT_ICON 1 resource with 3752 bytes
11744 INFO: Writing RT_ICON 2 resource with 2216 bytes
11744 INFO: Writing RT_ICON 3 resource with 1384 bytes
11744 INFO: Writing RT_ICON 4 resource with 37019 bytes
11744 INFO: Writing RT_ICON 5 resource with 9640 bytes
11744 INFO: Writing RT_ICON 6 resource with 4264 bytes
11744 INFO: Writing RT_ICON 7 resource with 1128 bytes
11788 INFO: Appending archive to EXE C:\Users\WNC.T1-1-1-Q-52026\Desktop\1018\build\AutomaticTestTool\AutomaticTestTool.exe
12297 INFO: Building EXE from EXE-00.toc completed successfully.
12304 INFO: checking COLLECT
12304 INFO: Building COLLECT because COLLECT-00.toc is non existent
12304 INFO: Building COLLECT COLLECT-00.toc
21584 INFO: Building COLLECT COLLECT-00.toc completed successfully.
### Steps to reproduce

PyInstaller: 4.5.1
Python: 3.9.1
Platform: Windows-10-10.0.19041-SP0(architecture : x64)
- opencv-python version
numpy                     1.21.3
opencv-python             4.5.4.58

##### Issue submission checklist

 - [v ] This is not a generic OpenCV usage question (looking for help for coding, other usage questions, homework etc.)
   
 - [ v] I have read the README of this repository and understand that this repository provides only an automated build toolchain for OpenCV Python packages (there is no actual OpenCV code here)
   
 - [ v] The issue is related to the build scripts in this repository, to the pre-built binaries or is a feature request (such as "please enable this additional dependency")
 - 

- [ v]  I'm using the latest version of ``opencv-python``

**Comment by clouds1203** [comment]
> It's already in use, from your logs above:
> 
> `8234 INFO: Loading module hook 'hook-cv2.py' from 'c:\users\wnc.t1-1-1-q-52026\appdata\local\programs\python\python39\lib\site-packages\_pyinstaller_hooks_contrib\hooks\stdhooks'...`
> 
> Someone needs to fix it because it doesn't work with the new opencv-python version. Please open an issue to the PyInstaller hook repository, your issue is not really related to this repository: https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues

Thank you!
It's a great help for me!

**Comment by alalek** [comment]
There is already related issue: https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/110

**Comment by alalek** [comment]
> which version of opencv-python is work well with pyinstaller?

Try previous release 4.5.3.56: https://pypi.org/project/opencv-python/#history

**Comment by asenyaev** [comment]
Hi @clouds1203!

Could you try to reproduce using the latest release of OpenCV Python packages - 4.5.4.60? Some issues with imports were fixed right there.

In addition, I couldn't reproduce this issue with the latest opencv-python version.

**Comment by hblab-haims** [comment]
In my case, Opencv-python version **4.5.4.60** solves this problem.

My PyInstaller: 4.7.

**Comment by asenyaev** [comment]
I've taken a deeper look at the problem and reproduced the issue with the latest release (4.5.4.60).

What I've done:

1. Firstly, I've installed `opencv-python` package (4.5.4.60 version), which has all components in `<path>\site-packages\cv2`, even `config.py` file is there.
```
python3 -m pip install -U opencv-python==4.5.4.60
```
2. As the next step I've installed `pyinstaller` (latest version):
```
python3 -m pip install -U pyinstaller
```
3. I've added test lines for custom app `myapp.py`:
```
import cv2

print(cv2.getBuildInformation())
```
4. Then I've run `pyinstaller` for my `myapp.py`:
```
python3 -m pyinstaller myapp.py
```
5. After building, I've run `myapp.exe` file.
6. Error from this issue has appeared:
```
ImportError: OpenCV loader: missing configuration file: ['config.py']
```

That's strange that I have `config.py` file in the `cv2` package, but I don't have it in my "app". I've printed [`LOADER_DIR` variable](https://github.com/opencv/opencv/blob/4.x/modules/python/package/cv2/__init__.py#L86) (added `print(LOADER_DIR)` in `<path>\site-packages\cv2\__init__.py`) to get the information where I am. So, it showed me the path not to the package, but to some temporary directory (e.g. `C:\Users\user\AppData\Local\Temp\_MEI21362\cv2`) which is created every time to run my "app". This directory has only `cv2.*.pyd` and `opencv_videoio_ffmpeg*.dll` files, but we hardly depends on `config*.py` files.

I guess to solve this issue, need to fix it [here](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-cv2.py) as @skvark mentioned before.

Another temporary solution which has worked for me is to define an absolute path to `cv2` package. For example (change `<path_to_python>`:
```
python3 -m pyinstaller -F myapp.py --paths="<path_to_python>\Lib\site-packages\cv2"
```

**Comment by asenyaev** [comment]
The [fix](https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/354) has been merged in [pyinstaller-hooks-contrib](https://github.com/pyinstaller/pyinstaller-hooks-contrib) repository. When the next release happens there, this problem will be fixed.

**Comment by asenyaev** [comment]
The release of `pyinstaller-hooks-contrib` has [happened](https://github.com/pyinstaller/pyinstaller-hooks-contrib/releases/tag/2021.4) today and I've tested it - the issue did not happen.

I've done the same pipeline as above, but also upgrading `pyinstaller` package with `--force-reinstall`. The steps:
1. Firstly, I've installed `opencv-python` package (4.5.4.60 version):
```
python3 -m pip install -U opencv-python==4.5.4.60
```
2. As the next step I've installed `pyinstaller` (latest version):
```
python3 -m pip install -U --force-reinstall pyinstaller
```
3. I've added test lines for custom app `myapp.py`:
```
import cv2

print(cv2.getBuildInformation())
```
4. Then I've run `pyinstaller` for my `myapp.py`:
```
python3 -m pyinstaller myapp.py
```
5. After building, I've run `myapp.exe` file.
6. I've seen a build information of opencv without any issues.

**Closing this issue.**
